gno.land/pkg/gnoland/app.go

```diff
@@ -106,8 +111,20 @@ func NewAppWithOptions(cfg *AppOptions) (abci.Application, error) {
 		},
 	)
 
+	// Set up the event collector
+	c := newCollector[validatorUpdate](
+		cfg.EventSwitch,      // global event switch filled by the node
+		validatorEventFilter, // filter fn that keeps the collector valid
+	)
```

Comment:
I don't understand the purpose of this generic-based intermediary collector. Isn't the filtering handler sufficient for the task at hand?

Reply:
I wanted to create a collector for any type of event (today it's one type of event, tomorrow it's gonna be something else) -- instead of making a custom collector every time, this one is generic and returns the specific type of event for which you've created it

To get events you need to do a couple of things:
- register a new listener to the event switch
- grab events and parse them -----> we use a custom filter to determine which events "stay"
- reset the event queue on each scrape (`getEvents`)

The filter in this scenario is just given an event and is meant to extract the relevant data, transforming it in the process if needed

Reply:
What's the value of creating new types and wrapping the filtering callback? I believe the filtering callback is the only necessary component, and the additional wrapping merely introduces indirection and unnecessary complexity.

Reply:
The filtering callback simply takes in an event and extracts data from it, but doesn't do the necessary steps like register to the event switch, keep the events and reset the event queue when the parsing is done - this is handled by the collector. The eventswitch we have in the SDK is not meant to be a store, but a notification system, hence why the [callbacks don't return anything](https://github.com/gnolang/gno/blob/f28444aea4f47508c42c5ba1a0069947d6ae9790/tm2/pkg/events/events.go#L43). The point I'm trying to make is that we need a "store" for these events that happen async regardless, and the collector serves this purpose

The non-generic approach that we can take is having the collector know how to filter the events without the callback, but that defeats the purpose of it, and requires us to write additional custom collectors every time

I really prefer to keep the collector logic in here and see how it will be used in the future. I was thinking we can possibly transition @gfanton's #2471 namespace usage (in the future) to use this event collector system (the /r/whatever/names emits an event for registration, the collector catches it and we update the VM store, afterwards we know to prevent `addpkg` deployments without calling the VM every time (but just lookup the go store)). The calls on #2471 are cheap, but they have overhead regardless

Reply:
> The non-generic approach that we can take is having the collector know how to filter the events without the callback, but that defeats the purpose of it, and requires us to write additional custom collectors every time

I don't get this part.

The collector who manages some buffering and cleaning is acceptable. However, the purpose of a pub-sub system is to provide pure filtering-based routing. N subscribers can specify filtering constraints to be called only when their filters match.

Or, if I understand correctly, then, you try to do everything here to manage a case where the collector needs to keep track of things from previous transactions. If that's the case, I think it's overkill and will anyway be non-deterministic, since it will depend on the node and when they start replaying the history from.

What about having the event system stateful within a single transaction, but not persisting across transactions? That way, we have determinism, and we just need a simple filtering-based routing system for the current transaction's events.

Reply:
If you think that my proposal cannot work and if you cannot find a simpler solution, I propose the following:

- Profit from the upcoming multinode experiment to identify new bugs
- Ensure we have multi-tx unit tests involving the collector and our gnosdk modules (open an issue)

Reply:
I agree, I'd like the event switch to be more smarter too, but right now it's just a simple "this happened, call something" approach in the SDK code

>The collector who manages some buffering and cleaning is acceptable. However, the purpose of a pub-sub system is to provide pure filtering-based routing. N subscribers can specify filtering constraints to be called only when their filters match.

This is exactly what the collector does now:
- you create the collector for a specific type of event:
    - the collector taps into the node's event system (event switch)
    - the collector registers each new event that happens on the node
    - the collector calls the callback it was created with (filter) to decide whether or not to keep the event
    - the collector returns the kept events upon a call (`getEvents()`)
- whatever logic you have asks the collector for any events that happened between scrapes:
    - the collector returns the kept events upon a call (`getEvents()`)

Reply:
@moul 

I've tried simplifying it to have a single collector with a filter map, but the SDK events do not expose a type (let alone a hashable one for map) 🙁 

I'm inclined to keep the current generic system because it keeps the code minimal, and doesn't require us to change aspects of SDK events -- this is one improvement area I plan to focus on in the upcoming period

I've opened up an issue for tracking much more robust app integration tests:
https://github.com/gnolang/gno/issues/2530